File: QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Jugador
    attr_accessor :casilla_actual, :encarcelado, :nombre, :carta_libertad, :saldo
    
    def initialize (nombre)
      @encarcelado = false
      @nombre = nombre
      @saldo = 0
      @casilla_actual = nil
      @carta_libertad = nil
      @propiedades = Array.new
    end
    
    def tengo_propiedades
      @propiedades.any?
    end
    
    def actualizar_posicion(casilla)
      if casilla.numero_casilla < @casilla_actual.numero_casilla
        modificar_saldo(Qytetet::SALDO_SALIDA)
      end
      
      @casilla_actual = casilla
      if casilla.soy_edificable
        if casilla.tengo_propietario && !casilla.propietario_encarcelado
          modificar_saldo(-casilla.cobrar_alquiler)
          return true
        end
        modificar_saldo(-casilla.coste) if casilla.tipo == TipoCasilla::IMPUESTO
      end
      false
    end
    
    def comprar_titulo
      puedo_comprar = false
      if(@casilla_actual.soy_edificable)
        tengo_propietario = @casilla_actual.tengo_propietario
        if !tengo_propietario
          coste_compra= @casilla_actual.coste
          if tengo_saldo(coste_compra)
            titulo=@casilla_actual.asignar_propietario(self)
            @propiedades << titulo
            modificar_saldo(-coste_compra)
            puedo_comprar = true
          end
        end
      end
      puedo_comprar
    end
    
    def devolver_carta_libertad
      devolver_carta = @carta_libertad
      @carta_libertad = nil
      devolver_carta
    end
    
    def ir_a_carcel(carcel)
      @casilla_actual = carcel
      @encarcelado = true
    end
    
    def modificar_saldo(cantidad)
      @saldo += cantidad
    end
    
    def obtener_capital
      capital = saldo
      @propiedades.each { |propiedad| capital += propiedad.casilla.coste + 
          (propiedad.casilla.num_casas + propiedad.casilla.num_hoteles)*propiedad.precio_edificar
        if propiedad.hipotecada
          capital -= propiedad.hipoteca_base
        end
      }
      capital
    end
    
    def obtener_propiedades_hipotecadas(hipotecada)
      @propiedades.select { |p| p.hipotecada == hipotecada }
    end
    
    def pagar_cobrar_por_casa_y_hotel(cantidad)
      @propiedades.each do |propiedad|
        total += cantidad*propiedad.casilla.num_hoteles + cantidad*propiedad.casilla.num_casas 
      end
      @saldo += total
    end
    
    def pagar_libertad(cantidad)
      libre = false
      if tengo_saldo(cantidad)
        modificar_saldo(-cantidad)
        libre = true
      end
      libre
    end
    
    def puedo_edificar_casa(casilla)
      casilla.se_puede_edificar_casa && @saldo > casilla.titulo_propiedad.precio_edificar
    end
    
    def puedo_edificar_hotel(casilla)
      casilla.se_puede_edificar_hotel && @saldo > casilla.titulo_propiedad.precio_edificar
    end
    
    def puedo_hipotecar(casilla)
      es_de_mi_propiedad(casilla)
    end
    
    def puedo_pagar_hipoteca(casilla)
      tengo_saldo(casilla.coste_hipoteca)
    end
    
    def puedo_vender_propiedad(casilla)
      es_de_mi_propiedad(casilla) && !casilla.esta_hipotecada
    end
    
    def tengo_carta_libertad
      @carta_libertad != nil
    end
    
    def vender_propiedad(casilla)
      modificar_saldo(casilla.vender_titulo)
      puts @propiedades.include?(casilla)
      eliminar_de_mis_propiedades(casilla)
    end
    
    private
    def cuantas_casas_hoteles_tengo
      total = 0
      @propiedades.each do |p|
        total += p.casilla.num_casas + p.casilla.num_hoteles
      end
      total
    end
    
    def eliminar_de_mis_propiedades(casilla)
      @propiedades.delete(casilla.titulo_propiedad)
      casilla.titulo_propiedad.propietario = nil
    end
    
    def es_de_mi_propiedad(casilla)
      @propiedades.include?(casilla.titulo_propiedad)
    end
    
    def tengo_saldo(cantidad)
      @saldo >= cantidad
    end
    
    def to_s
      "#{@nombre}"
    end
  end
end

File: QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author fjose
 */
public class Jugador {

   private boolean encarcelado;
   private String nombre;
   private int saldo;
   private Sorpresa cartaLibertad;
   private ArrayList<TituloPropiedad> propiedades;
   private Casilla casillaActual;
   static int factorEspeculador = 1;

   public Jugador(String nombre) {
      this.encarcelado = false;
      propiedades = new ArrayList<>();
      this.nombre = nombre;
      this.cartaLibertad = null;
      this.casillaActual = null;
      this.saldo = 0;
   }

   protected Jugador(Jugador jugador) {
      this.encarcelado = false;
      this.nombre = jugador.nombre;
      this.propiedades = jugador.propiedades;
      this.cartaLibertad = jugador.cartaLibertad;
      this.casillaActual = jugador.casillaActual;
      this.saldo = jugador.saldo;
      this.encarcelado = jugador.encarcelado;
   }

   protected void pagarImpuestos(int cantidad) {
      modificarSaldo(-cantidad);
   }

   protected Especulador convertirme(int fianza) {
      System.out.println("me convierto en especulador");
      return new Especulador(this, fianza);
   }

   public String getNombre() {
      return nombre;
   }

   public ArrayList<TituloPropiedad> getPropiedades() {
      return propiedades;
   }

   public int getSaldo() {
      return saldo;
   }

   protected boolean actualizarPosicion(Casilla casilla) {
      boolean tengoPropietario = false;
      if (casilla.getNumeroCasilla() < casillaActual.getNumeroCasilla()) {
         modificarSaldo(Qytetet.getInstance().SALDO_SALIDA);
      }

      setCasillaActual(casilla);

      if (casilla.soyEdificable()) {
         if (((Calle) casilla).tengoPropietario()) {
            tengoPropietario = true;
            if (!((Calle) casilla).propietarioEncarcelado()) {
               modificarSaldo(-((Calle) casilla).cobrarAlquiler());
            }
         }
      } else if (casilla.getTipo() == TipoCasilla.IMPUESTO) {
         pagarImpuestos(casilla.getCoste());
      }

      return tengoPropietario;
   }

   public boolean isEncarcelado() {
      return encarcelado;
   }

   void setEncarcelado(boolean encarcelado) {
      this.encarcelado = encarcelado;
   }

   public Casilla getCasillaActual() {
      return casillaActual;
   }

   void setCasillaActual(Casilla casillaActual) {
      this.casillaActual = casillaActual;
   }

   void setCartaLibertad(Sorpresa cartaLibertad) {
      this.cartaLibertad = cartaLibertad;
   }

   public boolean tengoPropiedades() {
      return !propiedades.isEmpty();
   }

   boolean comprarTitulo() {
      boolean puedoComprar = false;
      if (casillaActual.soyEdificable()) {
         if (!((Calle) casillaActual).tengoPropietario()) {
            if (tengoSaldo(casillaActual.getCoste())) {
               propiedades.add(((Calle) casillaActual).asignarPropietario(this));
               modificarSaldo(-casillaActual.getCoste());
               puedoComprar = true;
            }
         }
      }
      return puedoComprar;
   }

   Sorpresa devolverCartaLibertad() {
      Sorpresa Libertad;
      Libertad = cartaLibertad;
      cartaLibertad = null;
      return Libertad;
   }

   void irACarcel(Casilla casilla) {
      setCasillaActual(casilla);
      setEncarcelado(true);
   }

   void modificarSaldo(int cantidad) {
      saldo += cantidad;
   }

   ArrayList<TituloPropiedad> obtenerPropiedadesHipotecadas(boolean hipotecada) {
      ArrayList<TituloPropiedad> misPropiedades = new ArrayList();

      for (TituloPropiedad propiedad : this.propiedades) {
         if (propiedad.getHipotecada() == hipotecada) {
            misPropiedades.add(propiedad);
         }
      }
      return misPropiedades;
   }

   int obtenerCapital() {
      int PrecioInmuebles = 0;
      for (TituloPropiedad precio : propiedades) {
         if (precio.getHipotecada() == true) {
            PrecioInmuebles -= precio.getHipotecaBase();
         }
         PrecioInmuebles += precio.getPrecioEdificar() * ((Calle) precio.getCasilla()).getNumCasas();
         PrecioInmuebles += precio.getPrecioEdificar() * ((Calle) precio.getCasilla()).getNumHoteles();
         PrecioInmuebles += precio.getCasilla().getCoste();
      }
      return saldo + PrecioInmuebles;
   }

   void pagarCobrarPorCasaYHotel(int cantidad) {
      int numeroTotal = cuantasCasasHotelesTengo();
      modificarSaldo(cantidad * numeroTotal);
   }

   boolean pagarLibertad(int cantidad) {
      boolean tengoSaldo = tengoSaldo(cantidad);
      if (tengoSaldo) {
         modificarSaldo(-cantidad);
      }
      return tengoSaldo;
   }

   boolean puedoEdificar/*casa*/(Casilla casilla) {
      boolean esMia = esDeMiPropiedad(casilla);
      boolean tengoSaldo = false;

      if (esMia) {
         tengoSaldo = tengoSaldo(((Calle) casilla).getPrecioEdificar());
      }

      return esMia && tengoSaldo;
   }

//    boolean puedoEdificarHotel(Casilla casilla) {
//        
//    }
   boolean puedoHipotecar(Casilla casilla) {
      return esDeMiPropiedad(casilla);
   }

   boolean puedoPagarHipoteca(Casilla casilla) {
      return tengoSaldo((int) (1.1 * ((Calle) casilla).calcularValorHipoteca()));
   }

   boolean puedoVenderPropiedad(Casilla casilla) {
      boolean esMia = esDeMiPropiedad(casilla);
      boolean hipotecada = ((Calle) casilla).estaHipotecada();
      return esMia && !hipotecada;
   }

   public boolean tengoCartaLibertad() {
      return cartaLibertad != null;
   }

   void venderPropiedad(Casilla casilla) {
      int precioVenta = ((Calle) casilla).venderTitulo();
      modificarSaldo(precioVenta);
      eliminarDeMisPropiedades(casilla);
   }

   int cuantasCasasHotelesTengo() {
      int inmuebles = 0;
      for (TituloPropiedad inmueble : propiedades) {
         inmuebles += ((Calle) inmueble.getCasilla()).getNumCasas();
         inmuebles += ((Calle) inmueble.getCasilla()).getNumHoteles();

      }
      return inmuebles;
   }

   void eliminarDeMisPropiedades(Casilla casilla) {
      int numero = propiedades.indexOf(((Calle) casilla).getTitulo());
      propiedades.remove(numero);
   }

   boolean esDeMiPropiedad(Casilla casilla) {
      return propiedades.contains(((Calle) casilla).getTitulo());
   }

   boolean tengoSaldo(int cantidad) {
      boolean miSaldo = false;
      if (saldo >= cantidad) {
         miSaldo = true;
      }
      return miSaldo;
   }

   public int getFactorEspeculador() {
      return factorEspeculador;
   }

   @Override
   public String toString() {
      return "Nombre: " + nombre + "\nEncarcelado: " + encarcelado
          + "\nSaldo actual: " + saldo
          + "\n\n\tPropiedades obtenidas\n" + propiedades.toString();
   }
}
